core/compiler.common.jvm/src/org/jetbrains/kotlin/load/java/AbstractAnnotationTypeQualifierResolver.kt

```diff
@@ -0,0 +1,205 @@
+    }
+
+    private fun resolveJsr305CustomState(annotation: Annotation): ReportLevel? {
+        javaTypeEnhancementState.jsr305.userDefinedLevelForSpecificAnnotation[annotation.fqName]?.let { return it }
+        val enumValue = annotation.findAnnotation(MIGRATION_ANNOTATION_FQNAME)?.enumArguments(onlyValue = false)?.firstOrNull()
+            ?: return null
+        return javaTypeEnhancementState.jsr305.migrationLevel ?: when (enumValue) {
+            "STRICT" -> ReportLevel.STRICT
+            "WARN" -> ReportLevel.WARN
+            "IGNORE" -> ReportLevel.IGNORE
+            else -> null
+        }
+    }
+
+    private fun extractNullability(
+        annotation: Annotation, forceWarning: Annotation.() -> Boolean
+    ): NullabilityQualifierWithMigrationStatus? {
+        knownNullability(annotation, annotation.forceWarning())?.let { return it }
+
+        val typeQualifierAnnotation = resolveTypeQualifierAnnotation(annotation) ?: return null
+        val jsr305State = resolveJsr305AnnotationState(annotation)
+        if (jsr305State.isIgnore) return null
+        // TODO: the result of `forceWarning` will be overwritten - expected? Probably not.
+        return knownNullability(typeQualifierAnnotation, typeQualifierAnnotation.forceWarning())
+            ?.copy(isForWarningOnly = jsr305State.isWarning)
+    }
+
+    fun extractNullability(
+        annotations: Iterable<Annotation>, forceWarning: Annotation.() -> Boolean = { false }
+    ): NullabilityQualifierWithMigrationStatus? =
+        annotations.firstNotNullOfOrNull { extractNullability(it, forceWarning) }
+
+    fun extractMutability(annotations: Iterable<Annotation>): MutabilityQualifier? {
+        return annotations.fold(null as MutabilityQualifier?) { found, annotation ->
+            when (annotation.fqName) {
+                in READ_ONLY_ANNOTATIONS -> MutabilityQualifier.READ_ONLY
+                in MUTABLE_ANNOTATIONS -> MutabilityQualifier.MUTABLE
+                else -> found
+            }.also { if (found != null && found != it) return null /* inconsistent */ }
+        }
+    }
+
+    private fun extractDefaultQualifiers(annotation: Annotation): JavaDefaultQualifiers? {
+        resolveQualifierBuiltInDefaultAnnotation(annotation)?.let { return it }
+
+        val (typeQualifier, applicability) = resolveTypeQualifierDefaultAnnotation(annotation)
+            ?: return null
+        val jsr305State = resolveJsr305CustomState(annotation) ?: resolveJsr305AnnotationState(typeQualifier)
+        if (jsr305State.isIgnore) return null
+        // TODO: since we override the warning status, whether we force it in `extractNullability` is irrelevant.
+        //   However, this is probably not what was intended.
+        val nullabilityQualifier = extractNullability(typeQualifier) { false } ?: return null
+        return JavaDefaultQualifiers(nullabilityQualifier.copy(isForWarningOnly = jsr305State.isWarning), applicability)
```

Comment:
It seems there is a double copying of `nullabilityQualifier` with a specific `isForWarningOnly`: here and in `extractNullability`. Maybe remove `.copy(isForWarningOnly = jsr305State.isWarning)` knowing that `jsr305State.isWarning` has already been taken into account in `extractNullability`?

Reply:
I agree that this code is confusing. Initially I tried merging resolveJsr305CustomState and resolveJsr305AnnotationState, and this caused some tests to fail - but if you look really carefully, `jsr305State` here is of the TypeQualifierDefault-annotated class itself if specified, and only otherwise does it fall back to the type qualifier's report level:

    val jsr305State = resolveJsr305CustomState(annotation) ?: resolveJsr305AnnotationState(typeQualifier)

So the value here can be different.